lib/shared_components/button.dart

```diff
@@ -0,0 +1,69 @@
+import 'package:flutter/widgets.dart';
+import 'package:google_fonts/google_fonts.dart';
+
+// TODO: Move to appropiate file/folder
+final buttonStyle = GoogleFonts.getFont(
+  "IBM Plex Sans",
+  fontWeight: FontWeight.w600,
+  fontSize: 16,
+  color: const Color(0xFFFFFFFF),
+);
+
+// TODO: Move/Rename
+const decoration = BoxDecoration(
+  gradient: LinearGradient(begin: Alignment(1.00, -0.08), end: Alignment(-1, 0.08), colors: [Color(0xFF48A2AE), Color(0xFF005A78)]),
+);
+
+final _borderRadius = BorderRadius.circular(16);
+const _boxShadow = [
+  BoxShadow(
+    color: Color(0x26000000),
+    blurRadius: 15,
+    offset: Offset(0, 0),
+    spreadRadius: 0,
+  )
+];
+
+class Button extends StatelessWidget {
+  final String text;
+  final GestureTapCallback? onTap;
+  final BoxDecoration decoration;
+  final double width;
+  final double height;
+
+  const Button({
+    super.key,
+    required this.text,
+    this.onTap,
+    required this.decoration,
+    this.width = double.infinity,
+    this.height = 40,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    final buttonDecoration = decoration.copyWith(
+      borderRadius: _borderRadius,
+      boxShadow: _boxShadow,
+    );
+    final enabled = onTap != null;
+
+    return GestureDetector(
+      onTap: onTap,
+      child: MouseRegion(
+        cursor: enabled ? SystemMouseCursors.click : SystemMouseCursors.forbidden,
+        child: Container(
+          height: height,
+          width: width,
+          decoration: enabled
+              ? buttonDecoration
+              // TODO allow choosing the disabled styling (Also this doesn't like anything from the design)
+              : buttonDecoration.copyWith(color: const Color(0x3348A2AE)),
+          child: Center(
+            child: Text(text, style: buttonStyle),
+          ),
+        ),
+      ),
+    );
+  }
+}
```

Comment:
```suggestion
          child: ConstPart(text: text, buttonStyle: buttonStyle),
        ),
      ),
    );
  }
}

class ConstPart extends StatefulWidget {
  final String text;
  final TextStyle buttonStyle;

  const ConstPart({super.key, required this.text, required this.buttonStyle});
  
  @override
  State<StatefulWidget> createState() => _ConstPartState();
}

class _ConstPartState extends State<ConstPart> {
  late Widget textWidget = Text(widget.text, style: widget.buttonStyle);

  @override
  void didUpdateWidget(covariant ConstPart oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.text != oldWidget.text || widget.buttonStyle != oldWidget.buttonStyle) {
      debugPrint("Rendering new");
      textWidget = Text(widget.text, style: widget.buttonStyle);
    }
  }

  @override
  Widget build(BuildContext context) {
    debugPrint("Build called");
    return textWidget;
  }
}
```

When we create generic Widget’s I think it is a good exercise from the beginning we think in which parts are constant and which aren't.

Regarding buttons most often the text field given to `Text` widget would be constant at compile time. Because we inject text and style as variables we can’t define the `Text` widget as constant and it will be rebuilt on all rebuilds of the parent.

We could extract the constant part’s in a stateful widget and by overriding `didUpdateWidget` we only update a cached value of the widget on changes.

If you try above you see `debugPrint("Build called”)` is called whenever we use the toggle in terms and condition, but the `debugPrint("Rendering new”)` isn’t hit.

It is of course a balance since it adds additional complexity and in this example it would do much but if the caching was done “higher” in the three then more of the subtree could been avoided to be rebuilt.

Reply:
I don't really agree with this and it feels like a case of premature optimization.
There should be much simpler way to fix this, like having the constructor be constant.
I assume the issue is that the parent is being rebuilt here :thinking: